### ISSUE TITLE
Cannot Delete Sub-accounts & Users via Canvas API using API Key

### ISSUE THREAD

**Issue by CreepyGnome** [instructure/canvas-lms#1128]
It looks like I can do all the CRUD operations but delete with your API using an API key.

Is there any reason why delete was hidden behind a different authentication method that isn't supported by the API key system in Canvas?

Can you guys add support for this to also include the API Key?

It could be as simple as something like these:

DELETE /api/v1/accounts/:account_id/
DELETE /api/v1/users/:user_id/

It would be nice to be able to delete any account or user really since we can create them programmatically it would help in a number of cleanup operations that have to be done manually when it comes to account.

FYI... I found and tried your Delete a user from the root account:

DELETE /api/v1/accounts/:account_id/users/:user_id

This doesn't delete the user it just removes the login from the user but the user still remains.  So that appears to do the same thing as:

DELETE /api/v1/users/:user_id/logins/:id

**Comment by CreepyGnome** [comment]
As an update, the DELETE user from root account only really HIDEs the user from public listing but is still accessible via the direct user URL or via an API call to get details about the user.

All DELETEs should be real deletes that remove the data, otherwise, it should just be an update to deactivate or hide using PUT or PATCH not DELETE or POST.

I still haven't found a way to delete a sub-account via a public API.

**Comment by simonista** [comment]
Hey @CreepyGnome, you're right that it looks like we don't currently have an API to delete a subaccount.  I'm not sure whether that was just an oversight or if there was a deliberate reason. I'll try to ask around today and get back to you.

You're also right that Canvas takes the approach that most delete actions are really just "soft deletes", which update a state field on the record.  Usually the app is built so that you can't interact with these soft deleted records directly.  Looking at the endpoint you mentioned, it looks like it's mis-worded slightly, it doesn't actually mark the user record as deleted, just deletes all the associated enrollments and logins, effectively rendering the user orphaned.

You may want to look into the SIS import system: https://canvas.instructure.com/doc/api/file.sis_csv.html, it allows more control over user and sub account records.  Can you say more about what you're trying to accomplish?

**Comment by CreepyGnome** [comment]
Thanks @simonista for the response and I look forward to what you find out.

The soft deletes can become a problem for a number of reasons one, being the build of a large amount of dead data that is still in the database. However an even more critical one is legal requirements due to various regulations and certifications an entity using Canvas may be under that requires data after a specific age to be deleted, or that when specific types of data, like personal identifying information of students and teachers, are really deleted and not recoverable.

I am aware of the hacky workaround of exporting, then removing via CSV and then importing via your SIS features, however, that only works if you have an SIS connected. If you do not that export function always exports and empty file.  This is a poor workaround due to being very manual and not easy to automate programmatically the same way automation of the creation of data is done via the Canvas REST API.

**Comment by CreepyGnome** [comment]
Are there plans to add destroy APIs for all items stored DB and to make them cascading as an optional flag?  We are having trouble trying to actually remove courses as well, and all things related to them. Destroy versions of all the delete APIs would be awesome. Don't need UI features for the destroy APIs would be satisfactory.

**Comment by simonista** [comment]
No, there are currently no plans to add this.

**Comment by CreepyGnome** [comment]
Out of curiosity how Canvas help enable schools to easily stay FERPA compliant.

**Comment by simonista** [comment]
I don't think this is the right forum to discuss that, I'd encourage you to reach out to our product team in the canvas community: https://community.canvaslms.com/

**Comment by CreepyGnome** [comment]
Is that to say you have no documentation that explains how Canvas can help a school manager their FERPA requirements? 

I feel this is a valid place to bring up the issue, as this issue is directly tied to the need to delete/purge data from the database due to regulatory requirements that drove this issue we are chatting it to be created.

**Comment by omarkhan** [comment]
We use github issues to track bugs. As engineers we don't comment on FERPA compliance, that's a conversation you'll need to have with someone else.

**Comment by CreepyGnome** [comment]
Would you be able to provide me with contact details for who is specifically able to have that conversation or pass along mine to them (rsfoley@gmail.com)?  When I have asked questions on the Canvas LMS community site it has been hit or miss, and typically I get responses from non-Instrucuture employees.

**Comment by omarkhan** [comment]
Please reach out to your Instructure rep, the [mailing list](http://groups.google.com/group/canvas-lms-users), [IRC](http://webchat.freenode.net/?channels=canvas-lms), or the community site.

**Comment by CreepyGnome** [comment]
Seems odd to close this. You have added one destroy via commit 2593c83 seems logic that it would be reasonable to keep adding more over time.

I don't have an Instructure rep. We use the Open Source product. Im just trying to help improve Canvas when  I find issues.

I feel that a product manager should decide what issues are closed without completion seems odd to just close it the way it was as if I'm bothering you with bringing up issues in Canvas and trying to improve it for everyone.

**Comment by simonista** [comment]
It seems like there are really three requests mixed together in this issue:

1. Destroy sub-account was a missing api endpoint. That has been fixed.
2. Destroy user doesn’t actually mark the user as destroyed. Let’s file that as a new issue dedicated to just that. I agree that's probably an oversight.
3. Destroy does soft-delete, not hard-delete.  We have no current plans to implement this across all api endpoints. (although I will note that we have implemented it in one very specific instance where we felt it would be useful: https://canvas.instructure.com/doc/api/files.html#method.files.destroy) This would be considered a feature request, which we encourage everyone to submit to the Canvas Community: https://community.canvaslms.com/community/ideas. That's where it has the best chance of being prioritized by our product team.  We don't use GitHub issues as our project management tool.
